File: mipush_hook/src/main/java/Hook.java
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.startup.Initializer;

import com.xiaomi.channel.commonutils.android.DeviceInfo;
import com.xiaomi.channel.commonutils.android.MIUIUtils;
import com.xiaomi.smack.ConnectionConfiguration;

import java.lang.reflect.Field;
import java.util.Collections;
import java.util.List;

public class Hook implements Initializer<Void> {
    @NonNull
    @Override
    public Void create(@NonNull Context context) {
        try {
            doHook();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        return null;
    }

    @NonNull
    @Override
    public List<Class<? extends Initializer<?>>> dependencies() {
        return Collections.emptyList();
    }


    private void doHook() throws Exception {
        fakeMIUISystem();
        avoidTracking();
        ensureDefaultXmppServerIsCnHost();
    }

    private void fakeMIUISystem() throws Exception {
        hookField(MIUIUtils.class, "isMIUI", MIUIUtils.IS_MIUI);
    }

    private void avoidTracking() throws Exception {
        hookField(DeviceInfo.class, "sCachedIMEI", "");
    }

    private static void ensureDefaultXmppServerIsCnHost() {
        ConnectionConfiguration.setXmppServerHost(ConnectionConfiguration.XMPP_SERVER_CHINA_HOST_P);
    }

    private void hookField(Class klass, String field, Object value) throws Exception {
        Field target = klass.getDeclaredField(field);
        target.setAccessible(true);
        target.set(null, value);
    }
}

File: mipush_hook/src/main/java/Hook.kt
@file:Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
import android.content.Context
import androidx.startup.Initializer
import com.xiaomi.channel.commonutils.android.DeviceInfo
import com.xiaomi.channel.commonutils.android.MIUIUtils
import com.xiaomi.smack.ConnectionConfiguration
import java.lang.reflect.Field

class Hook : Initializer<Void> {
    override fun create(context: Context): Void {
        try {
            doHook()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
        @Suppress("UNCHECKED_CAST")
        return null as Void
    }

    override fun dependencies(): List<Class<out Initializer<*>>> = emptyList()

    @Throws(Exception::class)
    private fun doHook() {
        fakeMIUISystem()
        avoidTracking()
        ensureDefaultXmppServerIsCnHost()
    }

    @Throws(Exception::class)
    private fun fakeMIUISystem() {
        hookField(MIUIUtils::class.java, "isMIUI", MIUIUtils.IS_MIUI)
    }

    @Throws(Exception::class)
    private fun avoidTracking() {
        hookField(DeviceInfo::class.java, "sCachedIMEI", "")
    }

    private fun ensureDefaultXmppServerIsCnHost() {
        ConnectionConfiguration.setXmppServerHost(ConnectionConfiguration.XMPP_SERVER_CHINA_HOST_P)
    }

    @Throws(Exception::class)
    private fun hookField(klass: Class<*>, field: String, value: Any) {
        val target: Field = klass.getDeclaredField(field)
        target.isAccessible = true
        target.set(null, value)
    }
}
